File: Java/Arbitrary/runCollatz_arb.java
// Hannah Gulle
// Computes the top 10 collatz sequences by start and length for a 
// given range using arbitrary precision.
// Required classes: collatz_arb.java, sortByLength_arb.java, sortByStart_arb.java

import java.math.*;
import java.util.*;
import java.lang.Object;

// Main Program
public class runCollatz_arb{

	public static void main(String[] args){

		// The height of the sequence start values
		System.out.println("Input the Highest Starting Value");
		Scanner in = new Scanner(System.in);
		BigInteger end = in.nextBigInteger();

		// Holds all Collatz Objects found between 2 and the end
		ArrayList<collatz_arb> all = new ArrayList<collatz_arb>();		
		// Holds the elements of the current collatz sequence
		ArrayList<BigInteger> sequence = new ArrayList<BigInteger>();

		// Holds the size of the current collatz sequnce
		int size;
		BigInteger temp;
		for( int i = 2; i < end.intValue() ; i++ ){

			temp = BigInteger.valueOf(i);
		
			sequence = Collatz(temp);
			size = sequence.size()-1;

			all.add(new collatz_arb(temp, size));
		
			sequence.clear();		
		}

		System.out.println("Top 10 Starting Values After Sorting By Length");

		// Sort in ascending order by sequence length
		Collections.sort(all, new sortByLength_arb());	

		// Array List holds top 10 collatz objects
		ArrayList<collatz_arb> topTen = new ArrayList<collatz_arb>();
		int top = 1;
		for( int j = all.size()-1; j > 1; j-- ){

			if( all.get(j).getLength() > all.get(j-1).getLength() ){
				topTen.add( all.get(j) );
				System.out.println( (all.get(j).getStart()).toString() + " " + all.get(j).getLength() );
				top++;
			}
			if( top > 10 ){
				break;
			}
		}
		
		System.out.println("\nTop 10 Starting Values After Sorting by Start");		

		// Sort top 10 by starting value
		Collections.sort( topTen, new sortByStart_arb());

		// output top 10 after sorting by start
		for( int i = 9; i > -1; i-- ){
			System.out.println( topTen.get(i).getStart() + " " + topTen.get(i).getLength() );
		}
	}

	// Computes the collatz sequence for a given starting integer
	public static ArrayList<BigInteger> Collatz( BigInteger start ){
		
		ArrayList<BigInteger> sequence = new ArrayList<BigInteger>();
		sequence.add( start );

		BigInteger zero, one, two, three;
		one = BigInteger.valueOf(1); two = BigInteger.valueOf(2);
		three = BigInteger.valueOf(3); zero = BigInteger.valueOf(0);

		while( !start.equals( one ) ){

			if( start.mod( two ) == zero ){
				start = start.divide( two );
			}
			else{
				start = (start.multiply( three )).add( one );
			}
			sequence.add( start );
		}
		return sequence;
	}

}

File: Perl/Arbitrary/runCollatz_arb.pl
#!/usr/bin/perl

# Hannah Gulle
# Program computes the top 10 starting integers whose collatz
# sequences are the longest

# using collatz object with arbitrary precision
use collatz_arb;
use bignum;

print "Input Highest Starting Integer: ";
my $end = <stdin>;

# array to hold all collatz objects
my @all = ();
my $size = 0;

# array to hold top 10 collatz objects
my @topTen = ();

# retrieves the collatz seqeunce for each
# starting point within the range
for( my $i = 2; $i < $end; $i++ ){

	my @sequence = getCollatz( $i );
	$size = @sequence - 1;
	my $col = new collatz_arb( $i, $size );
	push @all, $col;
	@sequence = ();
}

# sorts all collatz objects by sequence size
print "\nTop 10 Starting Integers After Sorting by Size\n";
@all = sort {
	$a->getSize() <=> $b->getSize(); 
} @all;

# outputs the top 10 starting integers after sorting by size
# and adds each into the topTen array
$top = 1;
for( my $i = @all - 1; $i > 1; $i-- ){

	if( @all[$i]->getSize() > @all[$i-1]->getSize() ){
		push @topTen, @all[$i];
		print @all[$i]->getStart(). " " .@all[$i]->getSize(). "\n";
		$top++;
	}
	if( $top > 10 ){
		last;
	}
}

print "\nTop 10 Starting Integers After Sorting by Start\n";
# sorts top 10 collatz objects by sequence start
@topTen = sort {
	$a->getStart() <=> $b->getStart();
} @topTen;

# outputs top 10 after sorting by start
for( my $i = @topTen - 1; $i > -1; $i-- ){
	print @topTen[$i]->getStart()." ".@topTen[$i]->getSize()."\n";
}
# returns the collatz sequence for a given starting value
sub getCollatz {

	my $start = $_[0];
	my @sequence = ( $start );

	while( $start != 1 ){

		if( $start % 2 == 0 ){
			$start = $start / 2;
		}
		else{
			$start = (3 * $start) + 1;
		}
		push @sequence, $start;
	}
	return @sequence;
}
